ros2bag/ros2bag/verb/record.py

```diff
@@ -82,6 +82,16 @@ def add_arguments(self, parser, cli_name):  # noqa: D102
             '--compression-format', type=str, default='', choices=['zstd'],
             help='Specify the compression format/algorithm. Default is none.'
         )
+        parser.add_argument(
```

Comment:
For the queue size, if your CPU is capable of keeping up with the load, it should never have more than one message waiting on it; if any threads are available, one will immediately take a message when it's pushed to the queue, so the only time more than one message could be in the queue is if every thread is busy.  If the message load is fairly constant, that means the queue is going to consistently keep growing, so you will end up dropping messages no matter how large the queue is.  If message load is bursty, having a larger queue size could help to deal with bursts without losing anything, although there's not a good way to determine what an appropriate queue size is without profiling the specific system in question.

For the number of threads, what is optimal will also vary depending on the host system.  Since compression is largely CPU-bound, the number of concurrent threads supported by the underlying hardware is a good starting point, although fewer than that will probably be sufficient on a system with a lot of cores.

Still, it might be interesting to use the performance tests to demonstrate the general effectiveness here, so I'll have to take a look at that.

Reply:
Ok, here's some fun statistics.  First, I modified the `writer_benchmark` so that it can take options to enable compression and adjust the number of threads & queue size.  (@emersonknapp , let me know if you think that'd be more appropriate for a different PR -- I went ahead and included it here since it's for the purpose of benchmarking multi-threaded compression anyway)  I ran tests with no compression, compression with 1 thread, and compression with 16 threads, all using this set of options:
```bash
  -p frequency:=100 \
  -p size:=1000000 \
  -p instances:=16 \
  -p max_cache_size:=100
```

And I profiled them all using Intel VTune to compare them.  The first thing that I noticed is that because there's just a single thread pushing messages into `SequentialWriter::write`, that is a bottleneck in every case.  When no compression is enabled, the writing to storage is done on the same thread, and the latency there causes it to miss a lot of messages; my test reported 13,555 missed messages and only 15.2812% were successfully recorded.
![no-compression-summary](https://user-images.githubusercontent.com/12546134/96272698-1497c800-0f94-11eb-9d5e-c47b090cf1d3.png)
![no-compression-tree](https://user-images.githubusercontent.com/12546134/96272717-195c7c00-0f94-11eb-83c6-a384598c02fa.png)
Neither of the tests with compression enabled dropped any messages, probably because the actual compression & writing are all offloaded to other threads, even when it's set to only use 1 thread.  We can also see that the load of the single compression thread roughly matches the load of the main thread that's pushing to it.
![1-thread-summary](https://user-images.githubusercontent.com/12546134/96272739-20838a00-0f94-11eb-8eb5-b93baf512fd9.png)
![1-thread-tree](https://user-images.githubusercontent.com/12546134/96272748-22e5e400-0f94-11eb-8d3f-482bcb2f54df.png)
When 16 threads are running, the program doesn't actually finish any faster, since it still seems like it's bound by that main writer thread, but the load among the compressor threads looks like it's much more evenly balanced across cores.
![16-thread-summary](https://user-images.githubusercontent.com/12546134/96273296-c8995300-0f94-11eb-8512-63f53f022d77.png)
![16-thread-tree](https://user-images.githubusercontent.com/12546134/96273300-cafbad00-0f94-11eb-8c15-fbcd01f6ab31.png)

Based on the 16-thread profile, it looks like zstd's compression is fast enough that I'm not sure we can really overload it without also having multiple threads pushing messages at once, which would probably require a more significant redesign of `writer_benchmark`, although that probably also means that just a few compression threads would be more than sufficient on anything but the most heavily taxed systems.